### ISSUE TITLE
Caraya Discover Tests fails on internal error

### ISSUE THREAD

**Issue by AndrewHeim** [JKISoftware/Caraya#133]
Thanks for Caraya - it is great, and I'm taking time to post here to make a little bit better.

I have a large project (admittedly with some wild dependencies, but that's largely why we're working on it).  Whether I invoke Caraya's `Run Tests in Active Project `or its `Run Tests From Folder`, it dies.

Occasionally it returned an error from `Linker:Read Info From File`.  Eventually I dug in and found what was going on, in Discover Tests (in this case in `Test Runner.Project.lvclass:OnTestDiscovery.vi`, but the problem is general in each of those children).

![image](https://user-images.githubusercontent.com/10777768/121751034-b1324b80-caca-11eb-9230-b9a4c87dbd75.png)

When (not If, on a large enough project) a project has VIs with bad dependencies, that Linker call will return an error.  When it does, the whole Caraya engine goes kaput.  In the interest of moving quickly, I slapped a For loop around things and it runs:
![image](https://user-images.githubusercontent.com/10777768/121750520-c5297d80-cac9-11eb-8459-16a46419ebc2.png)

Why a For loop?  So that if a VI causes an error, we can discard it and still scan the rest of the project.  This is understandably slower, but is more robust - enough for me to load at least _some_ tests.  One better approach might be to call the Linker once per file and then combine the results (or refactor the search that's based on it being in that format).

The vexing thing about this is it's not a function of your test VIs.  It's a function of the _sum of all things in your project_.  So a bad VI in a Prototyping Bad Ideas folder can suddenly kill _all_ tests - and often without much meaningful feedback.

Again - I know this is all because my project's code is funky.  But it does run well for the customer, so it's not _that_ bad in the grand scheme of things.  If Caraya could return at least partial results in a case like this that would be desirable over crashing.  In an ideal world, my report would include a list of project items that caused problems from the Linker, along with the verbose output from that error, so I can go fix the problem.

Thanks!

**Comment by AndrewHeim** [comment]
Hmm it seems this problem is limited to _finding_ the tests, not to running the tests.  That gives us options for workarounds.  #58 seems promising, for instance.

**Comment by francois-normandin** [comment]
@AndrewHeim Thanks for the report. I think it describes well the issue.

If I understand the problem correctly, we need a more rugged way to discover tests such that a broken test VI does not stop the whole testing process, and you expect Caraya to report such failure of a test to run so you can pinpoint the error and correct it. 

Would you be able to provide a little test project containing a bad VI and a test runner launcher to show the failure at the Read Linker Info node? This would help greatly in patching this up.

It seems that you are presenting a valid use case that contradicts the reasoning I came up with during investigation of #101 .
In this thread, I argued that filtering bad tests at discovery time would result in a large overhead compared to reporting the failed tests at execution time. However, the test harness we developed was still finding the tests and didn't cause the linker info reader to fail. You have found a use case where not only the tests are "bad", but also that the linker info throws an error which prevents tests to be run.

From a preliminary inspection of the code, I think we could wrap this discovery method into a 2-step process where we first try it the way it is, and on failure, run the same input into a For Loop as you describe. We'd have to cache the errors in the Test Runner's private data and report the failure.

For this, I see these possible cases:
1. All VIs are testable, Read Linker Info does not generate any errors: no change to current behavior 
2. Some VIs are testable, Read Linker Info fails for some of them: Add "bad VIs" list to TestRunner, filter out error output so tests can proceed. Report list of bad VIs downstream within the Caraya Test Results. The "Discover Tests" method would not error out if at least one VI is testable.
3. All VIs are untestable (Read Linker Info reports errors for ALL paths): The "Discover Tests" method should report the error right away and prevent test execution.

Do you think the above-mentioned cases cover all our bases?

**Comment by francois-normandin** [comment]
The solution I'm leaning towards, as described above, will be part of release 1.2.
![image](https://user-images.githubusercontent.com/11728548/121804543-e3e85b00-cc14-11eb-9a14-632a476765d3.png)

However, there is currently an open question as to how **_to report this type of error._**.
I plan to run all tests and report Bad Linkage errors in the test result, like any normal test execution.
I'll explain the other possible report strategies and reasoning for rejecting them. If anyone has comments, you are more than welcome. It is important that this be done right.

IMO, there are three ways to report the error: 

### Inject the "Bad linkage" errors in the error wire AFTER all tests have executed.
As tests are discovered, those that generate bad linkage errors are set aside in the Test Runner's private data. The test are then executed as usual. After execution completes, the Caraya Interactive UI could report that all tests have passed (Test Results = PASS), but there would be a bad linkage error in the final error wire. CI Server would pick up the failed tests but have access only to the first linkage problem, since errors are merged in the error wire.

Conclusion: I believe this strategy should be rejected because it reports incomplete results, and multiple linkage errors would require multiple tests in-between each fix. This seems like a waste of time and not a desirable outcome. If an error is required, as a developer, I want to know all of the failed tests all once.

### No error being reported but keep linkage errors accessible to developer through an accessor.
This use case would be catering to developers who want to ignore bad linkage and skip them completely. The advantage is that a developer could see all tests as PASS in the project and old chunks of abandoned code would be skipped completely if they do not compile and are not called directly by the test execution. The huge caveat is that a standalone test that suddenly fails to compile would be filtered out and reported only through an accessor method which is not catchable by a CI server.

Conclusion: This scheme is potentially creating a situation where tests that work are suddenly ignored while they are still very valid tests. This cannot be chosen as a blanket default strategy. Developers interested in this scheme can override the Test Runner and catch/filter/report those errors by themselves through extension.

### Create a "Bad Linkage" section in the test report.
Discovered tests are executed and bad linkages are simply added in the report at the end of execution. Bad linkage errors are still maintained as a list in the Test Runner's private data, allowing developers to access the full list programmatically.
The result would show bad linkages as a test failure, without impeding all the other tests to run.

Conclusion: This is a strategy that is coherent with the framework's intended use, which is to quickly provide feedback on failures. It is also the only one that will play well with CI Servers and other automated processes.

**Comment by francois-normandin** [comment]
This would be the expected result in the interactive UI: 

![image](https://user-images.githubusercontent.com/11728548/121806846-860d4080-cc1f-11eb-88d4-125e981b09ca.png)

And in a report: 
`<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="onTestExecution" errors="0" skipped="0" tests="1" failures="0" timestamp="2021-06-13T08:14:51">
    <testcase classname="onTestExecution" name="VI not in an executable state (C:\Projects\JKI\Caraya\src\tests\test-runner\test-brokenTests\Broken Tests\Broken Test 0001.vi)"/>
  </testsuite>
  <testsuite name="Linkage Error" errors="0" skipped="0" tests="1" failures="0" timestamp="2021-06-13T08:14:51">
    <testcase classname="Linkage Error" name="Linkage error(s) captured"/>
  </testsuite>
    <testsuite name="Linkage Errors" errors="0" skipped="1" tests="1" failures="0" timestamp="2021-06-13T08:14:51">
      <testcase classname="Linkage Errors" name="Bad Linkage &quot;C:\Projects\JKI\Caraya\src\tests\test-runner\test-brokenTests\Broken Tests\MoreRecentLVVersion Test 0002.vi&quot;">
        <failure message="test failure">"FAIL"</failure>
      </testcase>
    </testsuite>
</testsuites>
`

**Comment by AndrewHeim** [comment]
I like your approach to the solution and your suggested fix.  I do think that deals with the possibilities as far as I can see them.

My one concern is that when run from the CLI, I want this situation (bad linkage) to trigger an error even if all of the executed tests pass.  That way something like GitLab won't give you a false "all clear" at the top level.

That may involve the addition of your first approach (Inject the "Bad linkage" errors in the error wire AFTER all tests have executed) in addition to the option you settled on (Create a "Bad Linkage" section in the test report").  

If we can resolve #130 so that the test report has good results, then we're good.  But if I get this fix without #130, then I'll be back in the same boat of GitLab thinking everything is great when really it's not.  

(Summary for those not familiar with #130 - GitLab and potentially other CI tools rely in part on a nonzero exit code returning from the CLI if any tests failed.  They do process the test results XML, but not to the extent of replacing that return code.  A recent change affected this behavior, and the test reports returned from Caraya's test runner VI at that point are empty, making it impossible for the CLI caller to inspect them.  So until then, we need Caraya to generate such an error for us.)

**Comment by AndrewHeim** [comment]
I'm glad you were able to create a bad linkage!  I can't send you this project, and wasn't sure how to recreate it outside of this project.

**Comment by francois-normandin** [comment]
About CLI behavior, yes the behavior as discussed in #130 is preserved in this fix.

I've added an "onLinkageErrors.vi" callback method that executes at the end of "Run Tests". (see the screenshot).
It takes any linkage errors and merges the errors in the output so that CLI (through Test Suite output) can report an error.
In parallel, it will add Linkage Errors to the report.

![image](https://user-images.githubusercontent.com/11728548/122041504-04481f00-cda7-11eb-8c6f-bfec8e724f05.png)

**Comment by AndrewHeim** [comment]
Ah that last snippet was the final thing I needed - now it's clear.  If there's a linkage problem, it appears as a failed test.  Everything else basically holds constant.  

I like it.

**Comment by francois-normandin** [comment]
Included in 1.2.0 to be released soon
https://github.com/JKISoftware/Caraya/milestone/6

**Comment by AndrewHeim** [comment]
I know you just closed this but I want to put a bug in your ear for another option should we need it.

My project still isn't making the linker happy, even though the tests do run (once we get over finding them).  The fix above kept Caraya from crashing, but didn't help much with finding the tests.  This, however crude, did:

![image](https://user-images.githubusercontent.com/10777768/122103475-e3e28980-cdd3-11eb-9779-8f3b12f92621.png)

The idea here was to search the nodes on the block diagram in a brute force way.  It's slower, yes.  The name filter for the nodes could be improved (but it's enough for where we're at now).  But as weird as it is that this and the linker _should_ return the same thing, the linker is more easily confused.  This is slower, but manages to find the tests.

I may have to hack this in to get this project to run, and I want to go back and fix whatever linking problems are there (though LV isn't giving me much work with in that direction).  But this is one possibility for another way to find tests.

Thanks again!
Andrew

**Comment by francois-normandin** [comment]
Case reopened @AndrewHeim

I think the difference is simply that your tests are in the dependencies, and not in the project.

Hint: 
![image](https://user-images.githubusercontent.com/11728548/122106067-750f2b80-cde7-11eb-8293-2fd8f0261a81.png)

**Comment by francois-normandin** [comment]
If this is confirmed, I would advocate that the Project Test Runner should not include a search in the dependencies, but only those VIs explicitly in the project. 
As an alternative for your project, you could copy the "Test Runner.Project" class and implement your method in the sibling class.

**Comment by AndrewHeim** [comment]
Alas, they are in the project.  This was some example code where I was getting the descendants of the Tests folder in the projects.  I recognize the ExcludeDependencies flag was false in my example, but I was manually checking the project and the results in this case.

But I like the suggestion of using another child class to maintain a local fork if I need it to load my tests.  Up to now it's been a disable structure in there instead.

At this point it's less effort to maintain a separate test discovery child than it is to fix this project, so I'll do that.  For now.  You've been very helpful and I don't want to inflict my project's dependency woes on you.  ;-)

**Comment by francois-normandin** [comment]
Your use case is important to cover, since it is likely representative of projects that developers inherit and would appreciate that Caraya tells them of problems in a way that is useful.
You "project's dependency woes", as you call them, are teaching us about Caraya's edge cases...

**Comment by francois-normandin** [comment]
Alright, I'll close this one and release 1.2.
If you get any new info that sheds light on what is happening, please re-open or create a new issue and reference this thread.
I'll proceed on the assumption that the workaround is acceptable for most people that would be in your situation (aka. create an override of the Test Runner class).
